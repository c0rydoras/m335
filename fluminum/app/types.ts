export type AngleUnit = "rad" | "deg" | "percent";
