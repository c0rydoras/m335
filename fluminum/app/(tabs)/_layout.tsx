import { Tabs } from "expo-router";
import React from "react";
import Icon from "~/components/icon";
import { Cable, EqualSquare, Magnet } from "lucide-react-native";

export default function TabLayout() {
  return (
    <Tabs
      screenOptions={{
        headerShown: false,
      }}
    >
      <Tabs.Screen
        name="index"
        options={{
          title: "Kabelfinder",
          tabBarIcon: (props) => <Icon icon={Cable} {...props} />,
        }}
      />
      <Tabs.Screen
        name="magnet-finder"
        options={{
          title: "Magnetfinder",
          tabBarIcon: (props) => <Icon icon={Magnet} {...props} />,
        }}
      />
      <Tabs.Screen
        name="spirit-level"
        options={{
          title: "Wasserwaage",
          tabBarIcon: (props) => <Icon icon={EqualSquare} {...props} />,
        }}
      />
    </Tabs>
  );
}
