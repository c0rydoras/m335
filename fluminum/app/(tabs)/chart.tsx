import {LineChart} from "react-native-chart-kit";
import {useEffect, useState} from "react";
import {Magnetometer} from "expo-sensors";
import FFT from "fft.js"

export default function chart(){

    const f = new FFT(1024)

    const [subscription, setSubscription] = useState(null);
    const [data, setDataNowWorking] = useState<{time:number,value:number}[]>([])

    const [displayData, setDisplayData] = useState([{
        time: 0,
        value: 0}
    ])

    const [timeStart,setTimeStart] = useState(0);
    const [timeEnd,setTimeEnd] = useState(0);

    useEffect(() => {
        Magnetometer.setUpdateInterval(1);
        const listener=Magnetometer.addListener((result)=>{
            setDataNowWorking((d)=>[{time: Date.now(),value:Math.sqrt( (result.x**2)+(result.y**2)+result.z**2)},...d])
            console.log(data)
            if(data.length == 1024){
                test()
            }
        })
        const callback = setInterval(()=>{
            /*setDisplayData(data.slice(-10))
            console.log(data)
            console.log(displayData)*/
        },1000)
        return ()=>{listener.remove()
            clearInterval(callback)}
    }, [data]);

    function test (){
        /*const storage = []
        console.log("moin")
        const inputData = f.toComplexArray(data.map((a)=>a.value),undefined)
        console.log("hi")*/
        let input = data.map(a=>a.value)
        let out = f.createComplexArray()
        f.realTransform(out,input)
        console.log(JSON.stringify( f.fromComplexArray(out,undefined)))
    }




return (
    <LineChart data={{
        labels: displayData.map(a=> a.time),
        datasets: [{
            data: displayData.map(a=>a.value),
            color: (opacity = 1) => `rgba(134, 65, 244, ${opacity})`
        },
            {
                data: [100],
                withDots: false
            }
        ]

    }

    } width={300} height={400} chartConfig={{backgroundColor:'red',color:(opacity = 1) => `rgba(134, 65, 244, ${opacity})`}}/>
)
}
