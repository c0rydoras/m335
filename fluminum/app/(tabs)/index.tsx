import { View } from "react-native";
import { Zap } from "lucide-react-native";
import { Text } from "~/components/ui/text";
import Icon from "~/components/icon";
import { useCallback, useEffect, useMemo, useState } from "react";
import { Magnetometer, MagnetometerMeasurement } from "expo-sensors";
import { useFocusEffect } from "expo-router";
import { Badge } from "~/components/ui/badge";
import * as Haptics from "expo-haptics";
import { useAtomValue } from "jotai";
import { feedbackValueAtom } from "../settings";
import { Audio } from "expo-av";
import { type Sound } from "expo-av/build/Audio";
import {CameraView} from "expo-camera";
import * as React from "react";

const foreground = "text-foreground fill-foreground";
const yellow = "text-yellow fill-yellow";

const calculateAbsoluteMagnetometerValue = (data: MagnetometerMeasurement) =>
  Math.sqrt(data.x ** 2 + data.y ** 2 + data.z ** 2);

export default function Screen() {
  const [absoluteMagnetometerValue, setAbsoluteMagnetometerValue] =
    useState<number>(0);

  const feedbackValue = useAtomValue<string[]>(feedbackValueAtom);
  // const soundRef = useRef<Sound | null>(null);

  const [isActive, setActive] = React.useState<boolean>(false)

  const [sound, setSound] = useState<Sound | null>(null);

  useFocusEffect(
    useCallback(() => {
      setActive(true)
      const listener = Magnetometer.addListener((data) => {
        setAbsoluteMagnetometerValue(calculateAbsoluteMagnetometerValue(data));
      });

      return () => {
        listener.remove();
        setActive(false)
      };
    }, []),
  );

 const foundCable = useMemo(
     ()=>absoluteMagnetometerValue > 200,
     [absoluteMagnetometerValue]
 )

  useEffect(() => {
    const loadSound = async () => {
      const { sound } = await Audio.Sound.createAsync(
        require("../../assets/geiger-sound.mp3"),
      );
      await sound.setIsLoopingAsync(true);
      setSound(sound);
    };
    loadSound();

    return () => {
      const unloadSound = async () => {
        await sound?.unloadAsync().catch(() => {});
      };
      unloadSound();
    };
  }, []);

  useEffect(() => {
    if (!foundCable||!isActive) {
      sound?.stopAsync().catch(() => {})
      return;
    }
    feedbackValue.includes("vibration") &&
      Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Heavy);
    feedbackValue.includes("audio") && sound?.playAsync().catch(() => {});
  }, [foundCable,isActive]);

  // NOTE: right now this doesn't actually find cables
  // TODO: this is temporary
  return (
    <View className="flex justify-center items-center gap-5 p-6 bg-secondary/3">
      <View className="h-[25%]">
        <Badge variant="destructive" className="rounded-md p-4">
          <Text className="text-2xl text-center">
            Es können nicht alle Stromführende Leitungen entdeckt werden!
          </Text>
        </Badge>
      </View>
      <View className="h-[75%]">
        <Icon
          icon={Zap}
          size={300}
          className={foundCable ? yellow : foreground}
        />
      </View>
      {feedbackValue.includes("visual") && <CameraView enableTorch={foundCable&&isActive} />}
    </View>
  );
}
