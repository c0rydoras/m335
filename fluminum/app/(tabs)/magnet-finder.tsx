import { Magnetometer } from "expo-sensors";
import * as React from "react";
import {View} from "react-native";
import { Text } from "~/components/ui/text";
import { AnimatedCircularProgress } from 'react-native-circular-progress';
import * as Haptics from 'expo-haptics';
import { Audio } from 'expo-av';
import {Sound} from "expo-av/build/Audio";

export default function Screen() {
  const [magnetometerValue, setMagnetometerValue] = React.useState<number>(0);
  const [sound, setSound] = React.useState<Sound>()

  async function loadSound() {
    const {sound} = await Audio.Sound.createAsync( require('../../assets/geiger-sound.mp3'));
    setSound(sound);
    await sound.setIsLoopingAsync(true);
    await sound.playAsync();
  }

  function calcSoundRate(absoluteMagnetometerValue:number) {
    return 1 + ((4*absoluteMagnetometerValue)/1000)
  }

  React.useEffect(() => {
    if (sound == undefined) return;

    let time = 0;
    let oldSoundRate = 1;
    const listener = Magnetometer.addListener(async (e) => {
        const absoluteMagnetometerValue = Number((Math.sqrt((e.x*e.x) + (e.y*e.y) + (e.z*e.z)).toFixed(2)));
        if (absoluteMagnetometerValue == 0) return;
        setMagnetometerValue(absoluteMagnetometerValue);
        if(Date.now() > time) {
            Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Heavy)
            const newSoundRate = calcSoundRate(absoluteMagnetometerValue)
            let soundRateUpdateThreshold = Math.abs((newSoundRate - oldSoundRate))
            if(soundRateUpdateThreshold < 0 || soundRateUpdateThreshold > 32) {
              soundRateUpdateThreshold = 1;
            }
            if(soundRateUpdateThreshold > 0.2) {
              sound.setRateAsync(newSoundRate, false);
              oldSoundRate = newSoundRate;
            }
            time = (Date.now() + (20000/absoluteMagnetometerValue))
        }
    });

    Magnetometer.setUpdateInterval(30);

    return () => {
      listener.remove();
      sound?.unloadAsync();
    };
  }, [sound]);

  React.useEffect(() => {
    loadSound();
  }, [])

  return (
        <View className="flex-1 justify-center items-center gap-5 p-6 bg-secondary/30">
          <AnimatedCircularProgress
              size={300}
              rotation={270}
              width={15}
              arcSweepAngle={180}
              fill={magnetometerValue/10}
              tintColor="#00e0ff"
              backgroundColor="#3d5875" />
          <Text className="text-muted-foreground">{magnetometerValue.toString()}</Text>
        </View>
  );
}
