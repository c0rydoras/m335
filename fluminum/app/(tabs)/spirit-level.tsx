import * as React from "react";
import { View } from "react-native";
import { Text } from "~/components/ui/text";

export default function Screen() {
  return (
    <View className="p-6 bg-secondary/30 relative h-72 flex gap-5 m-10">
      <View className="absolute top-0 inset-x-[50%]">
        <Text>Top</Text>
      </View>
      <View className="absolute right-0 inset-y-[50%]">
        <Text>Right</Text>
      </View>
      <View className="absolute left-0 inset-y-[50%]">
        <Text>Left</Text>
      </View>
      <View className="absolute bottom-0 inset-x-[50%]">
        <Text>Bottom</Text>
      </View>
    </View>
  );
}
