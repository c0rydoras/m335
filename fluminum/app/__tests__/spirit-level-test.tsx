import { calculatePoints } from "~/app/(tabs)/spirit-level";
import { calculateCorrectAngleUnit } from "~/app/(tabs)/spirit-level";
import { render } from "@testing-library/react-native";
import Svg, { Line, Polygon, Text } from "react-native-svg";
import { AngleUnit } from "../types";

describe("calculatePoints", () => {
  test("topBottomDeg 0 leftRightDeg 0", () => {
    const { pointsLeft, pointsTop } = calculatePoints(0, 0);
    expect(pointsLeft).toEqual("50,50 50,50 50,50");
    expect(pointsTop).toEqual("50,50 50,50 50,50");
  });
  test("topBottomDeg 45 leftRightDeg 0", () => {
    const { pointsLeft, pointsTop } = calculatePoints(45, 0);
    expect(pointsLeft).toEqual("50,50 50,50 50,50");
    expect(pointsTop).toEqual("50,50 25,25 75,25");
  });
  test("topBottomDeg 45 leftRightDeg 90", () => {
    const { pointsLeft, pointsTop } = calculatePoints(45, 90);
    expect(pointsLeft).toEqual("50,50 0,0 0,100");
    expect(pointsTop).toEqual("50,50 25,25 75,25");
  });
  test("topBottomDeg 0 leftRightDeg -90", () => {
    const { pointsLeft, pointsTop } = calculatePoints(0, -90);
    expect(pointsLeft).toEqual("50,50 100,100 100,0");
    expect(pointsTop).toEqual("50,50 50,50 50,50");
  });
});
describe("svgRendering", () => {
  test("svgRendering", () => {
    const tree = render(
      <Svg viewBox="0 0 100 100">
        <Polygon fill="#FFE500" points="50,50 0,0 0,100" />
        <Polygon fill="#FFE500" points="50,50 25,25 75,25" />
        <Line stroke={"black"} x1={0} y1={0} x2={100} y2={100} />
        <Line stroke={"black"} x1={100} y1={0} x2={0} y2={100} />
        <Text textAnchor="middle" y="20" x="50">
          {"90째"}
        </Text>
        <Text textAnchor="middle" y="80" x="50">
          {"-90째"}
        </Text>
        <Text textAnchor="middle" y="50" x="20">
          {"45째"}
        </Text>
        <Text textAnchor="middle" y="50" x="80">
          {"-45째"}
        </Text>
      </Svg>,
    );
    expect(tree).toMatchSnapshot();
  });
});
describe("calculate correct angle unit", () => {
  it.each([22, 100, 90, 47, 20, 100])(
    "should return the same value when calculating with rad",
    (radValue) => {
      expect(calculateCorrectAngleUnit("rad", radValue)).toEqual(radValue);
    },
  );
  it.each([
    { unit: "unit that doesnt exist", value: 22 },
    { unit: "some unfathomable unit", value: 100 },
    { unit: "asdf", value: 90 },
  ])(
    "should return the same value when calculating with non supported unit",
    (radValue) => {
      expect(
        calculateCorrectAngleUnit(radValue.unit as AngleUnit, radValue.value),
      ).toEqual(radValue.value);
    },
  );
  it.each([
    { value: Math.PI / 2, solution: 90 },
    { value: 1, solution: 57.2958 },
    { value: 3.5, solution: 200.535 },
    { value: 2 * Math.PI, solution: 360 },
    { value: 0, solution: 0 },
  ])("should correctly calculate degree when given rad", (radValue) => {
    expect(calculateCorrectAngleUnit("deg", radValue.value)).toBeCloseTo(
      radValue.solution,
    );
  });
  it.each([{ value: Math.PI / 4, solution: 100 }])(
    "should correctly calculate percent when given rad",
    (radValue) => {
      expect(calculateCorrectAngleUnit("percent", radValue.value)).toBeCloseTo(
        radValue.solution,
      );
    },
  );
});
