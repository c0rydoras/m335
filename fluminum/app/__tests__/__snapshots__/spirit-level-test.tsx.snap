// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svgRendering svgRendering 1`] = `
<RNSVGSvgView
  align="xMidYMid"
  bbHeight="100%"
  bbWidth="100%"
  focusable={false}
  meetOrSlice={0}
  minX={0}
  minY={0}
  style={
    [
      {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      {
        "flex": 0,
        "height": "100%",
        "width": "100%",
      },
    ]
  }
  vbHeight={100}
  vbWidth={100}
>
  <RNSVGGroup
    fill={
      {
        "payload": 4278190080,
        "type": 0,
      }
    }
  >
    <RNSVGPath
      d="M50 50 0 0 0 100z"
      fill={
        {
          "payload": 4294960384,
          "type": 0,
        }
      }
      propList={
        [
          "fill",
        ]
      }
    />
    <RNSVGPath
      d="M50 50 25 25 75 25z"
      fill={
        {
          "payload": 4294960384,
          "type": 0,
        }
      }
      propList={
        [
          "fill",
        ]
      }
    />
    <RNSVGLine
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      propList={
        [
          "stroke",
        ]
      }
      stroke={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      x1={0}
      x2={100}
      y1={0}
      y2={100}
    />
    <RNSVGLine
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      propList={
        [
          "stroke",
        ]
      }
      stroke={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      x1={100}
      x2={0}
      y1={0}
      y2={100}
    />
    <RNSVGText
      content={null}
      dx={[]}
      dy={[]}
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      font={
        {
          "textAnchor": "middle",
        }
      }
      rotate={[]}
      x={
        [
          "50",
        ]
      }
      y={
        [
          "20",
        ]
      }
    >
      <RNSVGTSpan
        content="90째"
        dx={[]}
        dy={[]}
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
        font={{}}
        rotate={[]}
        x={[]}
        y={[]}
      />
    </RNSVGText>
    <RNSVGText
      content={null}
      dx={[]}
      dy={[]}
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      font={
        {
          "textAnchor": "middle",
        }
      }
      rotate={[]}
      x={
        [
          "50",
        ]
      }
      y={
        [
          "80",
        ]
      }
    >
      <RNSVGTSpan
        content="-90째"
        dx={[]}
        dy={[]}
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
        font={{}}
        rotate={[]}
        x={[]}
        y={[]}
      />
    </RNSVGText>
    <RNSVGText
      content={null}
      dx={[]}
      dy={[]}
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      font={
        {
          "textAnchor": "middle",
        }
      }
      rotate={[]}
      x={
        [
          "20",
        ]
      }
      y={
        [
          "50",
        ]
      }
    >
      <RNSVGTSpan
        content="45째"
        dx={[]}
        dy={[]}
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
        font={{}}
        rotate={[]}
        x={[]}
        y={[]}
      />
    </RNSVGText>
    <RNSVGText
      content={null}
      dx={[]}
      dy={[]}
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      font={
        {
          "textAnchor": "middle",
        }
      }
      rotate={[]}
      x={
        [
          "80",
        ]
      }
      y={
        [
          "50",
        ]
      }
    >
      <RNSVGTSpan
        content="-45째"
        dx={[]}
        dy={[]}
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
        font={{}}
        rotate={[]}
        x={[]}
        y={[]}
      />
    </RNSVGText>
  </RNSVGGroup>
</RNSVGSvgView>
`;
